### PR TITLE
[SPARK-12481] [CORE] [STREAMING] [SQL] Remove usage of Hadoop deprecated APIs and reflection that supported 1.x

### PR DESCRIPTION
**Pull request by srowen**
Remove use of deprecated Hadoop APIs now that 2.2+ is required

### CODE REVIEW

**Comment by srowen** [comment]
Pinging folks like @vanzin @pwendell @rxin @sarutak 
I don't think this one is controversial, as it just follows from removing older Hadoop support and cleans up the build. Still, it's not a small change, so wouldn't mind any other set of eyes on it.

**Comment by rxin** [comment]
Yea I think this one is fine. 

(I haven't looked at the detailed changes though)

File: core/src/main/scala/org/apache/spark/SparkHadoopWriter.scala [hunk 1]
```diff
@@ -143,6 +141,12 @@ class SparkHadoopWriter(jobConf: JobConf)
     taskContext
   }
 
+  protected def newTaskAttemptContext(
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/SparkHadoopWriter.scala]: Is this still needed? It seems you replaced the call with a direct call to the constructor in a lot of places.

File: core/src/main/scala/org/apache/spark/rdd/NewHadoopRDD.scala [hunk 2]
```diff
@@ -26,11 +26,11 @@ import org.apache.hadoop.conf.{Configurable, Configuration}
 import org.apache.hadoop.io.Writable
 import org.apache.hadoop.mapreduce._
 import org.apache.hadoop.mapreduce.lib.input.{CombineFileSplit, FileSplit}
+import org.apache.hadoop.mapreduce.task.{TaskAttemptContextImpl, JobContextImpl}
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/rdd/NewHadoopRDD.scala]: nit: order

File: core/src/main/scala/org/apache/spark/scheduler/EventLoggingListener.scala [hunk 3]
```diff
@@ -97,7 +97,7 @@ private[spark] class EventLoggingListener(
    * Creates the log file in the configured log directory.
    */
   def start() {
-    if (!fileSystem.getFileStatus(new Path(logBaseDir)).isDir) {
+    if (!fileSystem.getFileStatus(new Path(logBaseDir)).isDirectory) {
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/scheduler/EventLoggingListener.scala]: Unrelated to this line: this class has the `hadoopFlushMethod` hack which probably can go away now, if you want to do more cleanup.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/SqlNewHadoopRDD.scala [hunk 4]
```diff
@@ -26,10 +26,10 @@ import org.apache.hadoop.conf.{Configurable, Configuration}
 import org.apache.hadoop.io.Writable
 import org.apache.hadoop.mapreduce._
 import org.apache.hadoop.mapreduce.lib.input.{CombineFileSplit, FileSplit}
+import org.apache.hadoop.mapreduce.task.{TaskAttemptContextImpl, JobContextImpl}
```
Comment (vanzin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/SqlNewHadoopRDD.scala]: nit: order

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/WriterContainer.scala [hunk 5]
```diff
@@ -24,10 +24,9 @@ import scala.collection.JavaConverters._
 import org.apache.hadoop.fs.Path
 import org.apache.hadoop.mapreduce._
 import org.apache.hadoop.mapreduce.lib.output.{FileOutputCommitter => MapReduceFileOutputCommitter}
+import org.apache.hadoop.mapreduce.task.TaskAttemptContextImpl
 import org.apache.spark._
```
Comment (vanzin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/WriterContainer.scala]: nit: not your fault but a good opportunity to add a blank line here.

**Comment by vanzin** [comment]
LGTM; there's at least one extra possible cleanup, but feel free to punt on that one.

File: core/src/main/scala/org/apache/spark/SparkHadoopWriter.scala (hunk 1 above)
Comment (srowen) [inline core/src/main/scala/org/apache/spark/SparkHadoopWriter.scala]: Yea, it's overridden in a mock object in `OutputCommitCoordinatorSuite` so I kept this

**Comment by srowen** [comment]
I'm pushing an update with those changes. I found a similar one in `FileBasedWriteAheadLogWriter` and one more in `FsHistoryProvider` regarding checking for safe mode.

BTW here are the additional cleanups that could be made if Hadoop 2.6+ was assumed:

- SparkHadoopUtil: `getFileSystemThreadStatistics`, `getFSBytesReadOnThreadCallback` and `getFSBytesWrittenOnThreadCallback` only actually work on Hadoop 2.5+
- Hadoop RDDs could always directly access SplitLocationInfo, InputSplitWithLocationInfo introduced in 2.5
- YARN integration tests could directly access tag-related APIs like ApplicationContext getApplicationTags from 2.4 (?)
- AM could cleanly handle `ApplicationAttemptNotFoundException`

Taken together those don't add up to much advantage, but it's not nothing. The bigger reason would be less JAR hell to deal with and being able to access new APIs more freely.

**Comment by srowen** [comment]
Merged to master

**Pull request closed without merging**